File: 1697. Checking Existence of Edge Length Limited Paths/Solution.java
class Solution {
    public boolean[] distanceLimitedPathsExist(int n, int[][] edgeList, int[][] queries) {
        UFDS uf = new UFDS(n);
        int q = queries.length, e = edgeList.length, j = 0;
        int[][] qIdx = new int[q][4];
        for (int i = 0; i < q; i++) {
            qIdx[i][0] = queries[i][0];
            qIdx[i][1] = queries[i][1];
            qIdx[i][2] = queries[i][2];
            qIdx[i][3] = i;
        }
        boolean[] res = new boolean[q];
        Arrays.sort(edgeList, (a, b) -> a[2] - b[2]);
        Arrays.sort(qIdx, (a, b) -> a[2] - b[2]);
        for (int[] query : qIdx) {
            while (j < e && edgeList[j][2] < query[2]) {
                uf.union(edgeList[j][0], edgeList[j][1]);
                j++;
            }
            res[query[3]] = uf.same(query[0], query[1]);
        }
        return res;
    }
}

class UFDS {
    int[] parent, rank;

    public UFDS(int n) {
        parent = new int[n];
        rank = new int[n];
        for (int i = 0; i < n; i++) parent[i] = i;
    }

    public int find(int x) {
        if (parent[x] != x) parent[x] = find(parent[x]);
        return parent[x];
    }

    public boolean same(int x, int y) {
        return find(x) == find(y);
    }

    public void union(int x, int y) {
        if (same(x, y)) return;
        int a = find(x), b = find(y);
        if (rank[a] > rank[b]) {
            parent[b] = a;
        } else {
            parent[a] = b;
            if (rank[a] == rank[b]) rank[b]++;
        }
    }
}

File: 1697. Checking Existence of Edge Length Limited Paths/Solution.py
class Solution:
    def distanceLimitedPathsExist(self, n: int, edgeList: List[List[int]], queries: List[List[int]]) -> List[bool]:
        q, e, j = len(queries), len(edgeList), 0
        uf = UFDS(n)
        edgeList.sort(key = itemgetter(2))
        qIdx = [[queries[i][0], queries[i][1], queries[i][2], i] for i in range(q)]
        qIdx.sort(key = itemgetter(2))
        res = [False for i in range(q)]
        for query in qIdx:
            while j < e and edgeList[j][2] < query[2]:
                uf.union(edgeList[j][0], edgeList[j][1])
                j += 1
            res[query[3]] = uf.same(query[0], query[1])
        return res

class UFDS:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.rank = [0 for i in range(n)]
    
    def find(self, x):
        if self.parent[x] != x: self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def same(self, x, y):
        return self.find(x) == self.find(y)
    
    def union(self, x, y):
        if self.same(x, y): return
        a, b = self.find(x), self.find(y)
        if self.rank[a] > self.rank[b]:
            self.parent[b] = a
        else:
            self.parent[a] = b
            if self.rank[a] == self.rank[b]: self.rank[b] += 1